File: dynbem/oye.py
"""Level 2 BEM with Øye 2-stage annular dynamic inflow.

Annulus-local alternative to Pitt-Peters.  Each radial annulus has its
own pair of first-order lag filters chasing the quasi-steady momentum
target; **no global L-matrix coupling** between annuli, no λ_c/λ_s
harmonic states.  Numerically much friendlier than Pitt-Peters at high
advance ratios and in descent + edgewise wind, where the
Pitt-Peters BEM-feedback (λ_c → BEM → C_L_hub → λ_s_ss) is stiff.

Inflow ODE (per annulus i, Øye 1990; OpenFAST AD Theory §6.3.4):

    τ₁ · dW_int/dt + W_int = W_qs + k · τ₁ · dW_qs/dt
    τ₂ · dW/dt     + W     = W_int

with coupling constant k = 0.6 (empirical, Øye original) and

    τ₁ = 1.1 / (1 - 1.3·min(a, 0.5)) · R / V_inf
    τ₂(r) = (0.39 - 0.26·(r/R)²) · τ₁

The blade sees W (not W_int, not W_qs).  This module's DBEMT_Mod
equivalent is OpenFAST's Mod=1: dW_qs/dt is treated as zero across the
outer integration step (constant τ assumption), which is exact for
operating-envelope sweeps and a small approximation for fast transients.

ψ-loop
------
Identical structure to PittPetersModelJIT._fwd_forces, except the local
inflow at each (r, ψ) is simply ``λ_climb + W[r]`` — no harmonic
decomposition.  Cyclic-frequency variation in blade loading still
emerges naturally because cyclic pitch and v_t_extra make local AoA
vary with ψ, and the hub moments Mx/My are still accumulated from the
azimuthally-varying dT.  But the *inflow itself* doesn't have explicit
cosine/sine harmonic states.

References
----------
Øye, S. (1990).  A simple vortex model.  IEA Symposium.
Snel, H. & Schepers, J.G. (1995).  Joint investigation of dynamic
  inflow effects and implementation of an engineering method.  ECN.
OpenFAST AeroDyn Theory v3.5, §6.3.4 (DBEMT).
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np
from numba import njit

from . import AeroBase, AeroResult, RotorInputs
from ._bem_common import (
    _interp_polar,
    build_polar_arrays,
    radial_grid,
    vrs_lambda1,
)
from .cyclic import cyclic_coeffs
from .rotor_definition import RotorDefinition
from .rotor_state import OyeRotorState, RotorState
from .polar import AirfoilPolar, LinearPolar, TabulatedPolar


# Øye coupling constant.  0.6 is the value Øye originally proposed and
# the OpenFAST default; it damps the coupling between W_qs and W_int.
_OYE_K = 0.6


# ---------------------------------------------------------------------------
# JIT'd ψ-loop kernel
# ---------------------------------------------------------------------------

@njit(cache=True, fastmath=True)
def _oye_psi_loop(r_arr, dr, chord, twist, col, omega, Omega_R,
                  lambda_climb, W, rho, n_b, n_psi,
                  v_in_hub_x, v_in_hub_y,
                  theta_1c, theta_1s,
                  alpha_tab, cl_tab, cd_tab,
                  out_dT_avg):
    """ψ-loop with per-annulus inflow W[r].

    Mirrors PittPetersModelJIT._fwd_forces but the local inflow comes
    from the annulus filter W[r] instead of harmonic decomposition.

    Out-parameter (pre-allocated by caller):
        out_dT_avg[i]  azimuth-averaged dT per annulus i (N)

    Returns (T_total, Q_total, Mx_hub, My_hub).  Each is azimuth-averaged
    (= 1/n_psi sum, matching the existing convention).
    """
    n_r = r_arr.shape[0]
    for i in range(n_r):
        out_dT_avg[i] = 0.0

    T_acc  = 0.0
    Q_acc  = 0.0
    Mx_acc = 0.0
    My_acc = 0.0
    inv_n_psi = 1.0 / n_psi

    for ipsi in range(n_psi):
        psi = 2.0 * math.pi * ipsi / n_psi
        cos_psi = math.cos(psi)
        sin_psi = math.sin(psi)
        v_t_extra = v_in_hub_x * sin_psi + v_in_hub_y * cos_psi
        col_psi = col + theta_1c * cos_psi + theta_1s * sin_psi
        rdT_sum = 0.0
        for i in range(n_r):
            r = r_arr[i]
            v_t = omega * r + v_t_extra
            if v_t <= 0.0:
                continue
            # Local inflow: uniform across annulus, no harmonics.
            lam_local = lambda_climb + W[i]
            v_a = lam_local * Omega_R
            phi = math.atan2(v_a, v_t)
            alpha = col_psi + twist - phi
            cl, cd = _interp_polar(alpha, alpha_tab, cl_tab, cd_tab)
            cos_p = math.cos(phi)
            sin_p = math.sin(phi)
            cn = cl * cos_p - cd * sin_p
            ct = cl * sin_p + cd * cos_p
            # No Prandtl tip loss in this JIT kernel — matches the
            # convention in pitt_peters_jit.  W_qs solver below uses
            # F=1 to stay consistent.
            q  = 0.5 * rho * (v_a*v_a + v_t*v_t) * chord * dr * n_b
            dT_elem = q * cn
            T_acc  += dT_elem
            Q_acc  += q * ct * r
            rdT_sum += r * dT_elem
            out_dT_avg[i] += dT_elem * inv_n_psi
        Mx_acc += rdT_sum * sin_psi
        My_acc += rdT_sum * cos_psi

    return T_acc * inv_n_psi, Q_acc * inv_n_psi, Mx_acc * inv_n_psi, My_acc * inv_n_psi


# ---------------------------------------------------------------------------
# Quasi-steady W_qs from per-annulus momentum balance
# ---------------------------------------------------------------------------

def _solve_W_qs(dT_avg: np.ndarray, x_arr: np.ndarray,
                dr: float, R: float, Omega_R: float, mu_T: float,
                rho: float) -> np.ndarray:
    """Quasi-steady induced inflow ratio per annulus, Glauert form.

    Annulus momentum with Glauert's resultant-velocity mass flow:

        dT = 4·π·r·dr·ρ·V_resultant·v_i

    where V_resultant = √(v_edge² + (v_climb + v_i)²) ≈ Ω·R·μ_T uses
    the rotor-mean μ_T (computed externally from the converged state).
    Linearised in W_qs, this gives the closed-form annulus solution

        W_qs[i] = dCT/dx[i] / (4·x[i]·μ_T)

    with dCT/dx = dT·R / (ρ·π·R²·Ω_R²·dr).  This matches the steady-
    state target Pitt-Peters uses (lam0_ss = T/(2ρA·V_T·Ω_R) in
    aggregate form) and stays stable in forward flight where the pure
    axial-momentum form  4·x·λ_r·W = dCT/dx  produces an enormous W
    when λ_r is small (the autorotation / VRS attractor).

    No Prandtl loss factor: matches the JIT ψ-loop convention.

    Returns W_qs array, same length as the radial grid.
    """
    if Omega_R < 1e-6:
        return np.zeros_like(dT_avg)
    A = math.pi * R * R
    rho_norm = rho * A * Omega_R * Omega_R * dr / R
    dCdx = dT_avg / max(rho_norm, 1e-30)
    W_qs = dCdx / (4.0 * np.maximum(x_arr, 1e-6) * max(mu_T, 0.05))
    # Sanity clamp — physical W is small; clip avoids NaN if a transient
    # state produces wild dT.
    return np.clip(W_qs, -1.0, 1.0)


def _oye_taus(R: float, x_arr: np.ndarray, V_inf: float, a_avg: float
              ) -> tuple[np.ndarray, np.ndarray]:
    """Øye time constants per annulus.

    τ₁ = 1.1 / (1 − 1.3·min(a_avg, 0.5)) · R / V_inf      (rotor-mean)
    τ₂(r) = (0.39 − 0.26·(r/R)²) · τ₁                      (radius-dependent)

    The 0.5 clamp on a_avg keeps τ₁ finite as induction approaches the
    actuator-disk limit a=0.5.  Matches OpenFAST DBEMT_Mod=1.
    """
    a_clamped = max(0.0, min(0.5, a_avg))
    tau1 = 1.1 / (1.0 - 1.3 * a_clamped) * R / max(V_inf, 1e-2)
    tau2_arr = (0.39 - 0.26 * x_arr * x_arr) * tau1
    tau1_arr = np.full_like(x_arr, tau1)
    return tau1_arr, tau2_arr


# ---------------------------------------------------------------------------
# Model
# ---------------------------------------------------------------------------

@dataclass
class OyeBEMModel(AeroBase):
    """Level 2 BEM with Øye 2-stage annular dynamic inflow.

    Stable alternative to PittPetersModel for high-advance-ratio and
    descent + edgewise operating regimes where Pitt-Peters' global
    L-matrix coupling can be numerically stiff.  Trade-off: doesn't
    capture cyclic-driven λ_c/λ_s harmonics explicitly — the inflow
    response to hub moments is purely through per-annulus loading
    asymmetry (which the ψ-loop captures), not through a coupled L
    matrix.
    """

    defn: RotorDefinition
    n_psi_elements: int = 36
    coupling_k:     float = _OYE_K
    _polar: AirfoilPolar = field(init=False, repr=False)

    def __post_init__(self) -> None:
        if self.defn.airfoil.polar_csv is not None:
            self._polar = TabulatedPolar.from_csv(self.defn.airfoil.polar_csv)
        else:
            self._polar = LinearPolar.from_properties(self.defn.airfoil)
        # Shared with PittPetersModel(JIT) — see dynbem/_bem_common.py.
        (self._dr, self._r_mid, self._x_mid, self._x_hub,
         self._twist_rad) = radial_grid(self.defn.blade)
        self._alpha_tab, self._cl_tab, self._cd_tab = build_polar_arrays(self._polar)
        self._n_elements = self.defn.blade.n_elements
        self._use_tip_loss = bool(self.defn.airfoil.tip_loss)

        # Scratch buffer for the JIT kernel — avoid per-call allocation.
        self._dT_avg_buf = np.zeros(self._n_elements, dtype=np.float64)

    # -----------------------------------------------------------------------

    def initial_rotor_state(self) -> OyeRotorState:
        return OyeRotorState.zeros(self._n_elements)

    def compute_forces(
        self,
        inputs: RotorInputs,
        state:  RotorState,
    ) -> tuple[AeroResult, RotorState]:
        assert isinstance(state, OyeRotorState)
        blade   = self.defn.blade
        omega   = state.omega_rad_s
        rho     = inputs.rho_kg_m3
        R       = blade.radius_m

        Omega_R = omega * R
        hub_axis = inputs.R_hub @ np.array([0.0, 0.0, 1.0])
        v_rel    = inputs.wind_world - inputs.v_hub_world
        v_climb  = float(np.dot(v_rel, hub_axis))
        v_inplane = v_rel - v_climb * hub_axis
        v_edge   = float(np.linalg.norm(v_inplane))
        mu       = v_edge / max(Omega_R, 1e-6)
        v_inplane_hub = inputs.R_hub.T @ v_inplane

        # Cyclic pitch coefficients.
        theta_1c, theta_1s = cyclic_coeffs(
            inputs.tilt_lon, inputs.tilt_lat, self.defn.control
        )
        has_cyclic = abs(theta_1c) + abs(theta_1s) > 1e-12

        # -------------------------------------------------------------------
        # Blade element forces — JIT ψ-loop with per-annulus W
        # -------------------------------------------------------------------
        T_total = Q_total = 0.0
        Mx_hub = My_hub = 0.0
        lambda_climb = v_climb / Omega_R if Omega_R > 1e-6 else 0.0

        if Omega_R > 1e-6 and omega > 1.0:
            T_total, Q_total, Mx_hub, My_hub = _oye_psi_loop(
                self._r_mid, self._dr,
                blade.chord_m, self._twist_rad,
                inputs.collective_rad, omega, Omega_R,
                lambda_climb, state.W, rho, blade.n_blades,
                self.n_psi_elements,
                float(v_inplane_hub[0]), float(v_inplane_hub[1]),
                theta_1c, theta_1s,
                self._alpha_tab, self._cl_tab, self._cd_tab,
                self._dT_avg_buf,
            )

        # -------------------------------------------------------------------
        # W_qs from per-annulus momentum balance (Glauert form), VRS override
        # -------------------------------------------------------------------
        # Rotor-mean μ_T = V_T / Ω·R, where V_T is the resultant velocity
        # through the disk.  Matches Pitt-Peters' aggregate denominator.
        v0_mean = float(np.mean(state.W)) * Omega_R
        V_T = max(math.sqrt(v_edge*v_edge + (v_climb + v0_mean)**2),
                  1e-2 * max(Omega_R, 1.0))
        mu_T = V_T / max(Omega_R, 1e-6)

        # Vortex Ring State: momentum theory diverges when 0 < V_descent/V_h < 2.
        # Use the Leishman empirical polynomial (same as PittPetersModel) and
        # apply it uniformly across annuli — Øye doesn't resolve VRS locally.
        rho_A = rho * math.pi * R * R
        T_pos = max(T_total, 0.0)
        V_h   = math.sqrt(T_pos / (2.0 * rho_A)) if T_pos > 1e-6 else 0.0
        V_c   = max(-v_climb, 0.0)
        lam2  = (V_c / V_h) if V_h > 1e-3 else 0.0
        if v_climb < -1e-3 and 0.0 < lam2 < 2.0 and Omega_R > 1e-6:
            W_uniform = vrs_lambda1(lam2) * V_h / Omega_R
            W_qs = np.full(self._n_elements, W_uniform)
        else:
            W_qs = _solve_W_qs(
                self._dT_avg_buf, self._x_mid,
                self._dr, R, Omega_R, mu_T, rho,
            )

        # -------------------------------------------------------------------
        # Øye 2-stage filter ODE
        # -------------------------------------------------------------------
        # τ₁, τ₂ use the same V_T (=mass-flow speed at the disk) and rotor-
        # mean a as the W_qs solver above.
        a_avg = float(np.mean(state.W) * Omega_R / V_T) if V_T > 1e-3 else 0.0
        tau1, tau2 = _oye_taus(R, self._x_mid, V_T, a_avg)

        # Mod=1: treat dW_qs/dt = 0 across the outer step (OpenFAST default
        # for envelope-style sweeps; the integrator's small dt makes this
        # exact for steady operating points).
        dW_int_dt = (W_qs - state.W_int) / tau1
        dW_dt     = (state.W_int - state.W) / tau2

        # -------------------------------------------------------------------
        # Mechanical states + assemble outputs
        # -------------------------------------------------------------------
        I_ode = (
            self.defn.autorotation.I_ode_kgm2
            if self.defn.autorotation.I_ode_kgm2 is not None
            else 1.0
        )
        d_omega = (-Q_total + inputs.motor_torque_Nm) / I_ode
        d_psi   = omega

        F_world   = -T_total * hub_axis
        M_orbital = inputs.R_hub @ np.array([Mx_hub, My_hub, 0.0])
        result = AeroResult(
            F_world=F_world,
            M_orbital=M_orbital,
            Q_spin=Q_total,
            M_spin=Q_total * hub_axis,
        )
        derivative = OyeRotorState(
            W_int=dW_int_dt,
            W=dW_dt,
            omega_rad_s=d_omega,
            spin_angle_rad=d_psi,
        )
        # Silence unused-variable lints — `mu`, `has_cyclic` are kept for
        # potential downstream gating but aren't currently used (Øye runs
        # the same ψ-loop in all regimes).
        _ = (mu, has_cyclic)
        return result, derivative

    # -----------------------------------------------------------------------

    def inflow_taus(self, inputs: RotorInputs, state: RotorState) -> np.ndarray:
        """Time constants for [W_int(n), W(n), ω, ψ].

        W_int uses τ₁, W uses τ₂(r); mechanical states get ∞.  Used by
        the envelope integrator's semi-implicit damping.
        """
        assert isinstance(state, OyeRotorState)
        omega = state.omega_rad_s
        R     = self.defn.blade.radius_m
        Omega_R = omega * R
        if Omega_R < 1e-6:
            n = self._n_elements
            return np.concatenate([
                np.full(n, np.inf), np.full(n, np.inf),
                [np.inf, np.inf],
            ])

        hub_axis = inputs.R_hub @ np.array([0.0, 0.0, 1.0])
        v_rel    = inputs.wind_world - inputs.v_hub_world
        v_climb  = float(np.dot(v_rel, hub_axis))
        v_edge   = float(np.linalg.norm(v_rel - v_climb * hub_axis))
        V_inf = max(math.sqrt(v_edge*v_edge + (v_climb + np.mean(state.W)*Omega_R)**2),
                    1e-2 * max(Omega_R, 1.0))
        a_avg = float(np.mean(state.W) * Omega_R / V_inf) if V_inf > 1e-3 else 0.0
        tau1, tau2 = _oye_taus(R, self._x_mid, V_inf, a_avg)
        return np.concatenate([tau1, tau2, [np.inf, np.inf]])

    def to_dict(self) -> dict:
        return {
            "model": "Oye_BEM_Level2",
            "n_elements": self.defn.blade.n_elements,
            "coupling_k": self.coupling_k,
        }

    @classmethod
    def from_definition(cls, defn: RotorDefinition) -> "OyeBEMModel":
        return cls(defn=defn)

File: dynbem_rs/src/oye.rs
// Level 2: Oye 2-stage annular dynamic inflow.
// See ../CLAUDE.md "Oye 2-stage annular dynamic inflow" and dynbem/oye.py.

use std::f64::consts::PI;

use crate::aero_io::{AeroResult, RotorInputs};
use crate::aero_model::AeroModel;
use crate::bem_common::{
    assemble_result, element_force, kinematics, v_t_disk, vrs_regime, ElementCtx, PsiKernel,
    RadialGrid, SweepCtx,
};
use crate::common::{
    vrs_lambda1, EPS_OMEGA_R, MIN_LOSS_FACTOR, VRS_DESCENT_THRESHOLD, V_T_HOVER_FLOOR_FRAC,
};
use crate::cyclic::cyclic_coeffs;
use crate::polar::Polar;
use crate::quasi_static_bem::{prandtl_hub_loss, prandtl_tip_loss};
use crate::rotor_definition::RotorDefinition;
use crate::rotor_state::OyeRotorState;

/// Oye empirical coupling constant (original Oye 1990, OpenFAST default).
/// Damps the coupling between the quasi-steady target W_qs and the
/// intermediate filter state W_int.
pub const OYE_K: f64 = 0.6;

/// Oye 2-stage filter step.
#[inline(always)]
pub fn oye_filter_step(
    d_w_int: &mut [f64],
    d_w: &mut [f64],
    w_qs: &[f64],
    w_int: &[f64],
    w: &[f64],
    tau2_arr: &[f64],
    tau1_scalar: f64,
    n_active: usize,
) {
    for i in 0..n_active {
        d_w_int[i] = (w_qs[i] - w_int[i]) / tau1_scalar;
        d_w[i] = (w_int[i] - w[i]) / tau2_arr[i];
    }
}

/// Upper clamp on rotor-mean induction factor `a` in the tau1 formula.
/// Without this, tau1 -> infinity as a -> 1/1.3 ~= 0.77 and the filter
/// stalls. 0.5 is the actuator-disk limit; matches OpenFAST DBEMT_Mod=1.
const A_AVG_CLAMP: f64 = 0.5;

/// Sanity clamp on quasi-steady inflow per annulus. Physical W is small
/// (~0.01-0.1); this catches transient blow-ups from a noisy dT estimate
/// without poisoning later steps.
const W_QS_CLAMP: f64 = 1.0;

/// Oye psi-loop kernel.
struct OyeKernel<'a> {
    lambda_climb: f64,
    w: &'a [f64],
    dt_avg: &'a mut [f64],
}

impl<'a> PsiKernel for OyeKernel<'a> {
    #[inline(always)]
    fn element<P: Polar>(&mut self, sweep: &SweepCtx<'_, P>, ctx: &ElementCtx) -> (f64, f64) {
        let lam = self.lambda_climb + self.w[ctx.i];
        let v_a = lam * sweep.omega_r;
        let (dt, dq) = element_force(v_a, sweep, ctx);
        self.dt_avg[ctx.i] += dt * sweep.n_psi_inv;
        (dt, dq)
    }
}

fn solve_w_qs(
    dt_avg: &[f64],
    x_arr: &[f64],
    f_loss: &[f64],
    n_active: usize,
    _dr: f64,
    r_tip: f64,
    omega_r: f64,
    mu_t: f64,
    rho: f64,
) -> Vec<f64> {
    if omega_r < EPS_OMEGA_R {
        return vec![0.0; n_active];
    }
    let area = PI * r_tip * r_tip;
    let rho_norm = (rho * area * omega_r * omega_r * _dr / r_tip).max(1e-30);
    let mu_t_safe = mu_t.max(crate::common::MU_T_FLOOR);
    let mut out = vec![0.0; n_active];
    for i in 0..n_active {
        let d_cdx = dt_avg[i] / rho_norm;
        let x = x_arr[i].max(EPS_OMEGA_R);
        let f = f_loss[i].max(MIN_LOSS_FACTOR);
        let w = d_cdx / (4.0 * x * f * mu_t_safe);
        out[i] = w.clamp(-W_QS_CLAMP, W_QS_CLAMP);
    }
    out
}

fn oye_taus(r_tip: f64, x_arr: &[f64], v_inf: f64, a_avg: f64, n_active: usize) -> (f64, Vec<f64>) {
    let a_c = a_avg.clamp(0.0, A_AVG_CLAMP);
    let tau1 = 1.1 / (1.0 - 1.3 * a_c) * r_tip / v_inf.max(V_T_HOVER_FLOOR_FRAC);
    let mut tau2 = vec![0.0; n_active];
    for i in 0..n_active {
        let x = x_arr[i];
        tau2[i] = (0.39 - 0.26 * x * x) * tau1;
    }
    (tau1, tau2)
}

#[derive(Clone)]
pub struct OyeBEMModel<P: Polar> {
    pub defn: RotorDefinition,
    pub n_psi_elements: usize,
    pub coupling_k: f64,
    pub polar: P,
    pub grid: RadialGrid,
}

impl<P: Polar + Clone> OyeBEMModel<P> {
    pub fn build(defn: RotorDefinition, n_psi_elements: usize, polar: P) -> Self {
        Self::build_with_k(defn, n_psi_elements, polar, OYE_K)
    }

    pub fn build_with_k(
        defn: RotorDefinition,
        n_psi_elements: usize,
        polar: P,
        coupling_k: f64,
    ) -> Self {
        let grid = RadialGrid::from_blade(&defn.blade);
        Self { defn, n_psi_elements, coupling_k, polar, grid }
    }
}

impl<P: Polar + Clone> AeroModel for OyeBEMModel<P> {
    type State = OyeRotorState;

    fn initial_state(&self) -> Self::State {
        let n = self.defn.blade.n_elements;
        OyeRotorState::zeros(n)
    }

    fn inflow_taus(&self, inputs: &RotorInputs, state: &Self::State) -> Vec<f64> {
        let r_tip = self.defn.blade.radius_m;
        let n = self.defn.blade.n_elements;
        let kin = kinematics(inputs, inputs.omega_rad_s, r_tip);
        let omega_r = kin.omega_r;
        if omega_r < EPS_OMEGA_R {
            return vec![f64::INFINITY; 2 * n];
        }
        let v_climb = kin.v_climb;
        let v_edge = kin.v_edge;
        let w_mean: f64 = if n > 0 {
            state.w_slice().iter().sum::<f64>() / (n as f64)
        } else {
            0.0
        };
        let v_inf = v_t_disk(v_edge, v_climb, w_mean * omega_r, omega_r);
        let a_avg = if v_inf > VRS_DESCENT_THRESHOLD {
            w_mean * omega_r / v_inf
        } else {
            0.0
        };
        let (tau1, tau2) = oye_taus(r_tip, &self.grid.x_mid[..n], v_inf, a_avg, n);
        let mut out = Vec::with_capacity(2 * n);
        out.extend(std::iter::repeat(tau1).take(n));
        out.extend(tau2);
        out
    }

    fn compute_forces(
        &self,
        inputs: &RotorInputs,
        state: &OyeRotorState,
    ) -> (AeroResult, OyeRotorState) {
        let blade = &self.defn.blade;
        let omega = inputs.omega_rad_s;
        let rho = inputs.rho_kg_m3;
        let r_tip = blade.radius_m;
        let n = blade.n_elements;

        let kin = kinematics(inputs, omega, r_tip);
        let omega_r = kin.omega_r;
        let hub_axis = kin.hub_axis;
        let v_climb = kin.v_climb;
        let v_edge = kin.v_edge;
        let v_inplane_hub = kin.v_inplane_hub;

        let gains = self.defn.control_gains();
        let (theta_1c, theta_1s) = cyclic_coeffs(inputs.tilt_lon, inputs.tilt_lat, gains);

        let lambda_climb = if omega_r > EPS_OMEGA_R {
            v_climb / omega_r
        } else {
            0.0
        };
        let mut dt_avg = vec![0.0; n];
        let (t_total, q_total, mx_hub, my_hub) = if omega_r > EPS_OMEGA_R && omega > 1.0 {
            let mut kernel = OyeKernel {
                lambda_climb,
                w: state.w_slice(),
                dt_avg: &mut dt_avg,
            };
            let sweep = SweepCtx {
                grid: &self.grid,
                polar: &self.polar,
                col: inputs.collective_rad,
                omega,
                omega_r,
                rho,
                n_b: blade.n_blades,
                n_psi: self.n_psi_elements,
                n_psi_inv: 1.0 / (self.n_psi_elements as f64),
                v_in_hub_x: v_inplane_hub[0],
                v_in_hub_y: v_inplane_hub[1],
                theta_1c,
                theta_1s,
            };
            sweep.run(&mut kernel)
        } else {
            (0.0, 0.0, 0.0, 0.0)
        };

        let w_mean: f64 = if n > 0 {
            state.w_slice().iter().sum::<f64>() / (n as f64)
        } else {
            0.0
        };
        let v0_mean = w_mean * omega_r;
        let vt_disk = v_t_disk(v_edge, v_climb, v0_mean, omega_r);
        let mu_t = vt_disk / omega_r.max(EPS_OMEGA_R);

        let area = PI * r_tip * r_tip;
        let vrs = vrs_regime(t_total, v_climb, rho, area);

        let use_tip_loss = self.defn.blade.tip_loss;
        let mut f_loss = vec![1.0; n];
        if use_tip_loss && omega_r > EPS_OMEGA_R {
            let n_b = self.defn.blade.n_blades;
            let x_hub = self.grid.x_hub;
            for i in 0..n {
                let r = self.grid.r_mid[i];
                let lam_local = lambda_climb + state.W[i];
                let v_a = lam_local * omega_r;
                let v_t = omega * r;
                let phi = v_a.atan2(v_t);
                let x = self.grid.x_mid[i];
                f_loss[i] = (prandtl_tip_loss(n_b, x, phi) * prandtl_hub_loss(n_b, x, x_hub, phi))
                    .max(MIN_LOSS_FACTOR);
            }
        }

        let w_qs: Vec<f64> = if vrs.in_vrs && omega_r > EPS_OMEGA_R {
            let w_uniform = vrs_lambda1(vrs.lam2) * vrs.v_h / omega_r;
            vec![w_uniform; n]
        } else {
            solve_w_qs(
                &dt_avg,
                &self.grid.x_mid[..n],
                &f_loss,
                n,
                self.grid.dr,
                r_tip,
                omega_r,
                mu_t,
                rho,
            )
        };

        let a_avg = if vt_disk > VRS_DESCENT_THRESHOLD {
            w_mean * omega_r / vt_disk
        } else {
            0.0
        };
        let (tau1_scalar, tau2_arr) = oye_taus(r_tip, &self.grid.x_mid[..n], vt_disk, a_avg, n);

        let mut d_w_int = vec![0.0; n];
        let mut d_w = vec![0.0; n];
        oye_filter_step(
            &mut d_w_int,
            &mut d_w,
            &w_qs,
            &state.W_int,
            &state.W,
            &tau2_arr,
            tau1_scalar,
            n,
        );

        let result = assemble_result(t_total, q_total, mx_hub, my_hub, hub_axis, &inputs.R_hub);
        let derivative = OyeRotorState {
            n_elements: n,
            W_int: d_w_int,
            W: d_w,
        };
        (result, derivative)
    }
}
